// @smartsheet-bridge/extension-handler/lib/models/BridgeChannelSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeChannelSettings = void 0;
class BridgeChannelSettings {
    constructor({ userId, threadId, runtimeData, } = {}) {
        if (userId)
            this.setUserID(userId);
        if (threadId)
            this.setThreadId(threadId);
        if (runtimeData)
            this.setRuntimeData(runtimeData);
    }
    static create(props = {}) {
        return new BridgeChannelSettings(props);
    }
    /**
     * Sets the unique identifier for the Bridge user.
     * @param threadId the unique identifier for the Bridge user.
     */
    setUserID(userId) {
        this.userId = userId;
    }
    /**
     * Sets the unique identifier for the Bridge thread.
     * @param threadId the unique identifier for the Bridge thread.
     */
    setThreadId(threadId) {
        this.threadId = threadId;
    }
    /**
     * Sets the user runtime data.
     *
     * Additional data that is passed to the next workflow but is not persisted.
     * @param runtimeData user runtime data.
     */
    setRuntimeData(runtimeData) {
        this.runtimeData = runtimeData;
    }
    toSerializableObject() {
        return {
            runtimeCTX: this.runtimeData,
            requestUUID: this.threadId,
            userUUID: this.userId,
        };
    }
}
exports.BridgeChannelSettings = BridgeChannelSettings;
