// @smartsheet-bridge/extension-handler/lib/models/ChannelUserInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelUserInfo = void 0;
class ChannelUserInfo {
    constructor({ email, firstName, lastName, locale, timezone, language, phoneNumber, } = {}) {
        if (email)
            this.setEmail(email);
        if (firstName)
            this.setFirstName(firstName);
        if (lastName)
            this.setLastName(lastName);
        if (locale)
            this.setLocale(locale);
        if (timezone)
            this.setTimezone(timezone);
        if (language)
            this.setLanguage(language);
        if (phoneNumber)
            this.setPhoneNumber(phoneNumber);
    }
    static create(props = {}) {
        return new ChannelUserInfo(props);
    }
    /**
     * Sets the user email address.
     * @param email email address.
     */
    setEmail(email) {
        this.email = email;
    }
    /**
     * Sets the first name of the user.
     * @param firstName user first name.
     */
    setFirstName(firstName) {
        this.firstName = firstName;
    }
    /**
     * Sets the last name of the user.
     * @param lastName user last name.
     */
    setLastName(lastName) {
        this.lastName = lastName;
    }
    /**
     * Sets the local of the user.
     * @param locale locale code.
     */
    setLocale(locale) {
        this.locale = locale;
    }
    /**
     * Sets the timezone of the user.
     * @param timezone timezone code.
     */
    setTimezone(timezone) {
        this.timezone = timezone;
    }
    /**
     * Sets the language of the user.
     * @param language language code.
     */
    setLanguage(language) {
        this.language = language;
    }
    /**
     * Sets the contact number of the user.
     * @param phoneNumber contact phone number.
     */
    setPhoneNumber(phoneNumber) {
        this.phoneNumber = phoneNumber;
    }
}
exports.ChannelUserInfo = ChannelUserInfo;
