// @smartsheet-bridge/extension-handler/lib/models/ChannelOutput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelOutput = void 0;
const handler_1 = require("@smartsheet-extensions/handler");
class ChannelOutput {
    constructor({ channelSetting, channelMessage, } = {}) {
        if (channelSetting)
            this.setChannelSetting(channelSetting);
        if (channelMessage)
            this.setChannelMessage(channelMessage);
    }
    static create(props = {}) {
        return new ChannelOutput(props);
    }
    /**
     * Sets the channel message.
     * @param channelMessage the channel message.
     */
    setChannelMessage(channelMessage) {
        this.channelMessage = channelMessage;
    }
    /**
     * Sets the channel settings.
     * @param channelSetting the channel settings.
     */
    setChannelSetting(channelSetting) {
        this.channelSetting = channelSetting;
    }
    toSerializableObject() {
        return {
            channelMessage: handler_1.serialize(this.channelMessage),
            channelSetting: handler_1.serialize(this.channelSetting),
        };
    }
}
exports.ChannelOutput = ChannelOutput;
