// @smartsheet-bridge/extension-handler/lib/models/MediaChannelMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaChannelMessage = void 0;
class MediaChannelMessage {
    constructor({ uid, type, data } = {}) {
        if (uid)
            this.setUid(uid);
        if (type)
            this.setType(type);
        if (data)
            this.setData(data);
    }
    static create(props = {}) {
        return new MediaChannelMessage(props);
    }
    /**
     * Sets the unique identifier for the channel message.
     * @param uid the unique identifier for the channel message.
     */
    setUid(uid) {
        this.uid = uid;
    }
    /**
     * Sets the rich media entity type.
     * @param type the rich media entity type.
     */
    setType(type) {
        this.type = type;
    }
    /**
     * Sets the rich media entity data.
     * @param data the rich media entity data.
     */
    setData(data) {
        this.data = data;
    }
    toSerializableObject() {
        return {
            uid: this.uid,
            media: {
                type: this.type,
                entity: this.data,
            },
        };
    }
}
exports.MediaChannelMessage = MediaChannelMessage;
