// @smartsheet-bridge/extension-handler/lib/models/TriggerWorkflowChannelMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerWorkflowChannelMessage = void 0;
class TriggerWorkflowChannelMessage {
    constructor({ uid, workflowID, stateValues, runtimeData, } = {}) {
        if (uid)
            this.uid = uid;
        if (workflowID)
            this.workflowID = workflowID;
        if (stateValues)
            this.stateValues = stateValues;
        if (runtimeData)
            this.runtimeData = runtimeData;
    }
    static create(props = {}) {
        return new TriggerWorkflowChannelMessage(props);
    }
    /**
     * Sets the unique identifier for the channel message.
     * @param uid the unique identifier for the channel message.
     */
    setUid(uid) {
        this.uid = uid;
    }
    /**
     * Set the identifier for the workflow to trigger.
     * @param workflowID the identifier for the workflow.
     */
    setWorkflowId(workflowID) {
        this.workflowID = workflowID;
    }
    /**
     * Sets the state values to set on the workflow run.
     * @param stateValues the state values to set.
     */
    setStateValues(stateValues) {
        this.stateValues = stateValues;
    }
    /**
     * Sets runtime data that will be passed to the workflow run when it is executed.
     * @param runtimeData the runtime data to send to the workflow run.
     */
    setRuntimeData(runtimeData) {
        this.runtimeData = runtimeData;
    }
    toSerializableObject() {
        return {
            uid: this.uid,
            conversation: {
                new: this.workflowID,
                entityData: this.stateValues,
                runtimeCTX: this.runtimeData,
            },
        };
    }
}
exports.TriggerWorkflowChannelMessage = TriggerWorkflowChannelMessage;
