// @smartsheet-bridge/extension-handler/lib/models/WorkflowTriggerSpec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowTriggerSpec = void 0;
class WorkflowTriggerSpec {
    constructor({ workflowID, stateValues, triggerDetails, } = {}) {
        if (workflowID)
            this.setWorkflowID(workflowID);
        if (stateValues)
            this.setStateValues(stateValues);
        if (triggerDetails)
            this.setTriggerDetails(triggerDetails);
    }
    static create(props = {}) {
        return new WorkflowTriggerSpec(props);
    }
    /**
     * Sets the workflow identifier.
     * @param workflowID the workflow identifier.
     */
    setWorkflowID(workflowID) {
        this.workflowID = workflowID;
    }
    /**
     * Sets any state values to be pre-set for the workflow.
     * @param stateValues pre-set state values.
     */
    setStateValues(stateValues) {
        this.stateValues = stateValues;
    }
    /**
     * Sets additional information about the triggering event.
     * @param triggerDetails additional information about the triggering event.
     */
    setTriggerDetails(triggerDetails) {
        this.triggerDetails = triggerDetails;
    }
    toSerializableObject() {
        return {
            intent: this.workflowID,
            entityData: this.stateValues,
            triggerDetails: this.triggerDetails,
        };
    }
}
exports.WorkflowTriggerSpec = WorkflowTriggerSpec;
