// @smartsheet-extensions/handler/lib/errors/AbstractError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractError = void 0;
const ExtensionResponse_1 = require("../responses/ExtensionResponse");
class AbstractError extends Error {
    constructor({ code, description, httpStatus, }) {
        super(description);
        Error.captureStackTrace(this, this.constructor);
        Object.defineProperty(this, 'name', {
            configurable: false,
            enumerable: false,
            writable: false,
            value: code || this.constructor.name,
        });
        Object.defineProperty(this, 'httpStatus', {
            configurable: false,
            enumerable: false,
            writable: false,
            value: httpStatus,
        });
    }
    toJSON() {
        return {
            status: ExtensionResponse_1.ExtensionStatus.FAIL,
            error: {
                code: this.name,
                httpStatus: this.httpStatus,
                description: this.message,
            },
        };
    }
}
exports.AbstractError = AbstractError;
