// @smartsheet-bridge/extension-handler/lib/responses/RegisterResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterResponse = void 0;
const handler_1 = require("@smartsheet-extensions/handler");
const WorkflowTriggerSpec_1 = require("../models/WorkflowTriggerSpec");
const AbstractResponse_1 = require("./AbstractResponse");
class RegisterResponse extends AbstractResponse_1.AbstractResponse {
    constructor({ settings, status, workflowTriggers, } = {}) {
        super(status);
        this.settings = settings;
        this.workflowTriggers = workflowTriggers;
    }
    static create(props = {}) {
        return new RegisterResponse(props);
    }
    /**
     * Set updated extension settings on the response.
     * @param settings the updated extension settings.
     */
    setSettings(settings) {
        this.settings = settings;
    }
    /**
     * Adds workflow trigger specs to the response.
     * @param triggers one or more workflow trigger specs.
     */
    addWorkflowTriggers(...triggers) {
        if (this.workflowTriggers === undefined) {
            this.workflowTriggers = [];
        }
        triggers.forEach(trigger => {
            this.workflowTriggers.push(WorkflowTriggerSpec_1.WorkflowTriggerSpec.create(trigger));
        });
    }
    toSerializableObject() {
        const { status } = handler_1.serializeObject(this);
        return {
            status,
            registrationData: handler_1.serialize(this.settings),
            workflowTriggers: handler_1.serialize(this.workflowTriggers),
        };
    }
}
exports.RegisterResponse = RegisterResponse;
