// @smartsheet-extensions/handler/lib/responses/ExtensionResponse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionStatus = void 0;
var ExtensionStatus;
(function (ExtensionStatus) {
    /**
     * The function execution was a success.
     */
    ExtensionStatus[ExtensionStatus["SUCCESS"] = 0] = "SUCCESS";
    /**
     * The function execution has failed.
     * */
    ExtensionStatus[ExtensionStatus["FAIL"] = 1] = "FAIL";
    /**
     * The function execution should be retried.
     * */
    ExtensionStatus[ExtensionStatus["RETRY"] = 2] = "RETRY";
    /**
     * Stop the current conversation flow and any parent conversations.
     * */
    ExtensionStatus[ExtensionStatus["STOP"] = 3] = "STOP";
    /**
     * Pause the current conversation flow, this will block the conversation until it is restarted.
     * */
    ExtensionStatus[ExtensionStatus["PAUSE"] = 4] = "PAUSE";
    /**
     * Pause the current conversation flow and start the OAuth authentication flow. The conversation will be restarted when authentication is successful.
     * */
    ExtensionStatus[ExtensionStatus["NEED_AUTH"] = 5] = "NEED_AUTH";
    /**
     * End the current conversation flow, will allow parent conversations to continue.
     * */
    ExtensionStatus[ExtensionStatus["END"] = 6] = "END";
})(ExtensionStatus = exports.ExtensionStatus || (exports.ExtensionStatus = {}));
