// @smartsheet-bridge/extension-cli-logger/lib/formats/tty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tty = void 0;
const chalk_1 = require("chalk");
const triple_beam_1 = require("triple-beam");
const winston_1 = require("winston");
const ASCII_ARROW = '➤';
const ASCII_ERROR = '‼︎';
const ASCII_WARN = '!';
const ASCII_BOX_TOP = '┌';
const ASCII_BOX_MIDDLE = '│';
const ASCII_BOX_BOTTOM = '└';
const Symbols = {
    error: ASCII_ERROR,
    warn: ASCII_WARN,
    info: ASCII_ARROW,
    verbose: ASCII_ARROW,
    debug: ASCII_ARROW,
};
const Colors = {
    error: chalk_1.hex('#E57373'),
    warn: chalk_1.hex('#FFF176'),
    info: chalk_1.blue,
    verbose: chalk_1.blue,
    debug: chalk_1.reset,
};
let isGrouped = false;
exports.tty = winston_1.format(info => {
    const msgs = [Colors[info.level](Symbols[info.level])];
    if (info.grouped === true) {
        msgs.push(ASCII_BOX_TOP);
        isGrouped = true;
    }
    else if (info.grouped === false) {
        msgs.push(ASCII_BOX_BOTTOM);
        isGrouped = false;
    }
    else if (isGrouped) {
        msgs.push(ASCII_BOX_MIDDLE);
    }
    msgs.push(info.message);
    return Object.assign(Object.assign({}, info), { [triple_beam_1.MESSAGE]: msgs.join(' ') });
});
