// @smartsheet-bridge/extension-handler/lib/handlers/handleOAuth2HandleCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleOAuth2HandleCode = exports.OAUTH2_HANDLE_CODE = void 0;
const handler_1 = require("@smartsheet-extensions/handler");
const BadHandleOAuth2CodeResponseError_1 = require("../errors/BadHandleOAuth2CodeResponseError");
const OAuth2Data_1 = require("../models/OAuth2Data");
const HandleOAuth2CodeResponse_1 = require("../responses/HandleOAuth2CodeResponse");
exports.OAUTH2_HANDLE_CODE = 'OAUTH2_HANDLE_CODE';
const isHandleOAuth2CodePayload = (payload) => payload.event === exports.OAUTH2_HANDLE_CODE;
const isOAuth2Data = (o) => {
    return handler_1.isSerializableObject(o) && o.access_token !== undefined;
};
exports.handleOAuth2HandleCode = (config) => create => () => {
    const next = create();
    return (body, callback) => {
        if (typeof config.onOAuthHandleCode !== 'function') {
            throw new handler_1.BadRequestError('onOAuthHandleCode function has not been defined');
        }
        if (!isHandleOAuth2CodePayload(body)) {
            throw new handler_1.BadRequestError('Payload must contain `event` property and `payload` property.');
        }
        const settings = (body.payload && body.payload.registrationData) || {};
        const { caller } = body;
        const { providerOAuth } = body.payload;
        next(config.onOAuthHandleCode({
            code: body.payload.code,
            state: body.payload.state,
            scope: body.payload.scope,
            oauthType: body.payload.oauthType,
        }, {
            settings,
            caller,
            redirectURI: body.payload.redirectURI,
            oAuthData: providerOAuth,
        }), (err, result) => {
            if (err) {
                callback(err);
            }
            else if (result instanceof HandleOAuth2CodeResponse_1.HandleOAuth2CodeResponse) {
                callback(err, result);
            }
            else if (result instanceof OAuth2Data_1.OAuth2Data) {
                callback(err, HandleOAuth2CodeResponse_1.HandleOAuth2CodeResponse.create(result));
            }
            else if (isOAuth2Data(result)) {
                callback(err, HandleOAuth2CodeResponse_1.HandleOAuth2CodeResponse.create(OAuth2Data_1.OAuth2Data.create(result)));
            }
            else if (result === null) {
                throw new BadHandleOAuth2CodeResponseError_1.BadHandleOAuth2CodeResponseError('null');
            }
            else {
                throw new BadHandleOAuth2CodeResponseError_1.BadHandleOAuth2CodeResponseError(typeof result);
            }
        });
    };
};
