// @smartsheet-extensions/handler/lib/utils/serializable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeObject = exports.serializeArray = exports.serialize = exports.isSerializableClass = exports.isSerializable = exports.isSerializableObject = exports.isSerializableArray = exports.isSerializablePrimitive = exports.isSerializableEmpty = void 0;
exports.isSerializableEmpty = (o) => o === null || o === undefined;
exports.isSerializablePrimitive = (o) => typeof o === 'string' || typeof o === 'number' || typeof o === 'boolean';
exports.isSerializableArray = (o) => Array.isArray(o) && o.every(exports.isSerializable);
exports.isSerializableObject = (o) => typeof o !== 'function' &&
    !exports.isSerializableEmpty(o) &&
    !exports.isSerializablePrimitive(o) &&
    !exports.isSerializableArray(o) &&
    Object.values(o).every(exports.isSerializable);
exports.isSerializable = (o) => exports.isSerializableEmpty(o) ||
    exports.isSerializablePrimitive(o) ||
    exports.isSerializableArray(o) ||
    exports.isSerializableObject(o);
exports.isSerializableClass = (o) => typeof o === 'object' &&
    typeof o.toSerializableObject === 'function';
exports.serialize = (o) => {
    if (exports.isSerializablePrimitive(o) || exports.isSerializableEmpty(o)) {
        return o;
    }
    if (exports.isSerializableClass(o)) {
        return o.toSerializableObject();
    }
    if (Array.isArray(o)) {
        return exports.serializeArray(o);
    }
    if (typeof o === 'object' && !Array.isArray(o) && typeof o !== 'function') {
        return exports.serializeObject(o);
    }
};
exports.serializeArray = (a) => a.map(exports.serialize).filter(e => e !== undefined);
exports.serializeObject = (o) => Object.entries(o).reduce((json, [key, value]) => (Object.assign(Object.assign({}, json), { [key]: exports.serialize(value) })), {});
